### PR TITLE
Add config for Doxygen and build instructions.

### PR DESCRIPTION
**Pull request by cmarqu**
References #774 (but doesn't close it).

### CODE REVIEW

File: documentation/source/library_reference_c.rst [hunk 1]
```diff
@@ -0,0 +1,21 @@
+########################
+C Code Library Reference
+########################
+
+Cocotb contains a library called ``GPI`` (in directory :file:`cocotb/lib/gpi/`) written in C
```
Comment (imphil) [inline documentation/source/library_reference_c.rst]: nit: It's actually mostly C++

File: documentation/Doxyfile [hunk 2]
```diff
@@ -0,0 +1,2427 @@
...
+SHORT_NAMES            = NO
+
+# If the JAVADOC_AUTOBRIEF tag is set to YES then doxygen will interpret the
+# first line (until the first dot) of a Javadoc-style comment as the brief
+# description. If set to NO, the Javadoc-style will behave just like regular Qt-
+# style comments (thus requiring an explicit @brief command for a brief
+# description.)
+# The default value is: NO.
+
+JAVADOC_AUTOBRIEF      = NO
```
Comment (imphil) [inline documentation/Doxyfile]: Please set this to yes, it's common style.

File: documentation/Doxyfile [hunk 3]
```diff
@@ -0,0 +1,2427 @@
...
+
+TCL_SUBST              =
+
+# Set the OPTIMIZE_OUTPUT_FOR_C tag to YES if your project consists of C sources
+# only. Doxygen will then generate output that is more tailored for C. For
+# instance, some of the names that are used will be different. The list of all
+# members will be omitted, etc.
+# The default value is: NO.
+
+OPTIMIZE_OUTPUT_FOR_C  = YES
```
Comment (imphil) [inline documentation/Doxyfile]: We have a lot of C++ code, please set this to NO.

File: documentation/Doxyfile [hunk 4]
```diff
@@ -0,0 +1,2427 @@
...
+
+# If you use STL classes (i.e. std::string, std::vector, etc.) but do not want
+# to include (a tag file for) the STL sources as input, then you should set this
+# tag to YES in order to let doxygen match functions declarations and
+# definitions whose arguments contain STL classes (e.g. func(std::string);
+# versus func(std::string) {}). This also make the inheritance and collaboration
+# diagrams that involve STL classes more complete and accurate.
+# The default value is: NO.
+
+BUILTIN_STL_SUPPORT    = NO
```
Comment (imphil) [inline documentation/Doxyfile]: You can set this to YES.

**Comment by cmarqu** [comment]
All done now.

File: documentation/Doxyfile (hunk 4 above)
Comment (cmarqu) [reply documentation/Doxyfile]: This actually broke breathe and in turn Sphinx, so I have left this ``NO``.

File: documentation/source/conf.py [hunk 5]
```diff
@@ -267,5 +266,24 @@
...
-# For now show the todoy 
+# For now show the todos
 todo_include_todos = True
+
+# -- Extra setup for C documentation with Doxygen and breathe 
+# see also https://breathe.readthedocs.io/en/latest/readthedocs.html
+
+read_the_docs_build = os.environ.get('READTHEDOCS', None) == 'True'
+if read_the_docs_build:
+    subprocess.call('cd ..; doxygen', shell=True)
```
Comment (imphil) [inline documentation/source/conf.py]: `subproces.call` has a `cwd` argument which is more suited than `cd ..`. And you don't need `shell` in this case.

File: documentation/source/conf.py [hunk 6]
```diff
@@ -267,5 +266,24 @@
...
+    subprocess.call('cd ..; breathe-apidoc -o source/generated source/doxygen/_xml -f', shell=True)
+else:  # local
+    subprocess.call('cd ..; doxygen', shell=True)
+    subprocess.call('cd ..; .venv/bin/breathe-apidoc -o source/generated source/doxygen/_xml -f', shell=True)
```
Comment (imphil) [inline documentation/source/conf.py]: How about adding `.venv/bin` to `$PATH` and  using the same code path for both build types? You can do that by passing `env=` to `subprocess.call`. And if you pass the individual arguments as list you can avoid `shell=True` as well.

File: documentation/source/library_reference_c.rst [hunk 7]
```diff
@@ -0,0 +1,21 @@
+########################
+C Code Library Reference
+########################
+
+Cocotb contains a library called ``GPI`` (in directory :file:`cocotb/lib/gpi/`) written in C++
```
Comment (imphil) [inline documentation/source/library_reference_c.rst]: It's in `cocotb/share/lib/` now.

File: documentation/source/library_reference_c.rst [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+that is an abstraction layer for the VPI, VHPI, and FLI simulator interfaces.
+
+.. image:: diagrams/svg/cocotb_overview.svg
+
+The interaction between Python and GPI is via a Python extension module called ``simulator``
+(in directory :file:`cocotb/lib/simulator/`) which provides routines for
```
Comment (imphil) [inline documentation/source/library_reference_c.rst]: `cocotb/share/lib/simulator/`

**Comment by cmarqu** [comment]
Okay, I resolved these,  but I'm now getting 
https://github.com/michaeljones/breathe/issues/431 in my RTD builds, and I'm running out of time to investigate this. I'll try to look at it further, but it may not happen soon.

**Comment by cmarqu** [comment]
Oh, okay, requiring Sphinx>=2.0 fixed it. Are we ready to do that step?

**Comment by imphil** [comment]
> Oh, okay, requiring Sphinx>=2.0 fixed it. Are we ready to do that step?

sure, that's fine.

I just gave your branch a try and now I get `Duplicate declaration` warnings for all C/C++ functions, e.g. 

```
/home/philipp/src/cocotb/documentation/source/generated/file/gpi__priv_8h.rst:4: WARNING: Duplicate declaration, int gpi_register_impl(GpiImplInterface  * func_tbl)
```

Does this come from reading c and h files? Or is there something else that needs to be fixed?

**Comment by cmarqu** [comment]
Yeah, I think I read back then that this is not easily fixed. If the result looks fine it's okay for now I guess?

**Comment by cmarqu** [comment]
For some reason Github still thinks there are unresolved review items, but there are none I'd say...

**Comment by imphil** [comment]
Looks good. Can you combine the three commits into one and rebase on top of master? Thanks!

**Pull request merged**